README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# openmetrics <img src='man/figures/logo.png' align="right" height="139" />

<!-- badges: start -->

[![travis-ci build
status](https://travis-ci.org/atheriel/openmetrics.svg?branch=master)](https://travis-ci.org/atheriel/openmetrics)
[![CRAN
status](https://www.r-pkg.org/badges/version/openmetrics)](https://cran.r-project.org/package=openmetrics)
<!-- badges: end -->

**openmetrics** is an opinionated client for
[Prometheus](https://prometheus.io/) and the related
[OpenMetrics](https://openmetrics.io/) project. It makes it possible to
add predefined and custom metrics to any R web application and expose
them on a `/metrics` endpoint, where they can be consumed by Prometheus
services.

The package includes built-in support for
[Plumber](https://www.rplumber.io/) and
[Shiny](https://shiny.rstudio.com/) applications, but is highly
extensible.

## Installation

You can install **openmetrics** from CRAN with

``` r
install.packages("openmetrics")
```

or from GitHub with

``` r
# install.packages("remotes")
remotes::install_github("atheriel/openmetrics")
```

## Use with Plumber

You can easily wrap an existing Plumber router object to add metrics for
HTTP requests and their duration:

``` r
srv <- plumber::plumb("plumber.R")
srv <- register_plumber_metrics(srv)
srv$run()
```

This will automatically create a `/metrics` endpoint that exposes these
metrics (and any others you have defined).

### Authentication

To add authentication to this endpoint, set the
`METRICS_HTTP_AUTHORIZATION` environment variable to the expected
[`Authorization`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization)
header you want Prometheus to use. For example, to grant access to the
username “aladdin” with password
    “opensesame”:

    Sys.setenv(METRICS_HTTP_AUTHORIZATION = "Basic YWxhZGRpbjpvcGVuc2VzYW1l")

Meanwhile, in your Prometheus configuration:

``` yaml
scrape_configs:
# ...
- job_name: my-plumber-api
  basic_auth:
    username: aladdin
    password: opensesame
  # ...
```

## Use with Shiny

You can also wrap an existing Shiny app object to add metrics on
reactive flush duration and a running session count. In your `app.R`
file, you can use something like the following:

``` r
app <- shiny::shinyApp(...)
app <- register_shiny_metrics(app)
app
```

Which will be picked up by `shiny::runApp()`.

Again, this will automatically create a `/metrics` endpoint that exposes
these metrics (and any others you have defined). It supports the same
authentication method as well.

**This feature should be considered experimental**. It relies on certain
unstable Shiny internals to add the `/metrics` endpoint, which [is not
usually possible](https://github.com/rstudio/shiny/issues/2395).

## Default Process Metrics

You can enable the [default process
metrics](https://prometheus.io/docs/instrumenting/writing_clientlibs/#process-metrics)
(which track CPU time, memory usage, and open files) with

``` r
register_default_metrics()
```

Note that not all metrics are supported (or even meaningful) on all
operating systems.

## Custom Metrics

All required metrics and features of [the informal Prometheus client
specification](https://prometheus.io/docs/instrumenting/writing_clientlibs/)
are supported, so you can create app-specific [counters, gauges, and
histograms](https://prometheus.io/docs/concepts/metric_types/). If you
want to make use of labels, you must pass a default value at the time of
creation.

Here are some examples collected from around the
house:

``` r
meows <- counter_metric("meows", "Heard around the house.", labels = "cat")
meows$inc(cat = "Shamus") # Count one meow from Shamus.
meows$inc(3, cat = "Unknown") # Count three meows of unknown origin.

thermostat <- gauge_metric("thermostat", "Thermostat display.")
thermostat$set(21.3) # Read from the display...
thermostat$dec(2) # ... and then turn it down 2 degrees.

temperature <- histogram_metric(
  "temperature", "Ambient room temperature measurements.",
  buckets = c(10, 15, 20, 22, 25), room = "kitchen"
)
set.seed(9090)
# Simulate taking ambient temperature samples.
for (measure in rnorm(20, mean = 21.5)) {
  temperature$observe(measure, room = sample(c("kitchen", "bathroom"), 1))
}
```

All metrics (in fact all exported functions) take a `registry` parameter
in case you want to avoid using the default global registry in some part
of your application. You can construct new registries with `registry()`.

### Performance

Because metrics may be gathered with high frequency, some effort has
been made to ensure that using them is fast – usually around 10
microseconds.

### Rendering Metrics Manually

The `render_metrics()` function exposes the [text-based OpenMetrics
format](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md)
understood by Prometheus. Here’s what the metrics above look like:

``` r
cat(render_metrics())
```

    # HELP meows Heard around the house.
    # TYPE meows counter
    meows_total{cat="Shamus"} 1
    meows_created{cat="Shamus"} 1604597246.05814
    meows_total{cat="Unknown"} 3
    meows_created{cat="Unknown"} 1604597246.05893
    # HELP thermostat Thermostat display.
    # TYPE thermostat gauge
    thermostat 19.3
    # HELP temperature Ambient room temperature measurements.
    # TYPE temperature histogram
    temperature_bucket{room="bathroom",le="10.0"} 0
    temperature_bucket{room="bathroom",le="15.0"} 0
    temperature_bucket{room="bathroom",le="20.0"} 0
    temperature_bucket{room="bathroom",le="22.0"} 9
    temperature_bucket{room="bathroom",le="25.0"} 11
    temperature_bucket{room="bathroom",le="+Inf"} 11
    temperature_sum{room="bathroom"} 234.387663039796
    temperature_count{room="bathroom"} 11
    temperature_created{room="bathroom"} 1604597246.08967
    temperature_bucket{room="kitchen",le="10.0"} 0
    temperature_bucket{room="kitchen",le="15.0"} 0
    temperature_bucket{room="kitchen",le="20.0"} 1
    temperature_bucket{room="kitchen",le="22.0"} 4
    temperature_bucket{room="kitchen",le="25.0"} 9
    temperature_bucket{room="kitchen",le="+Inf"} 9
    temperature_sum{room="kitchen"} 198.854388891071
    temperature_count{room="kitchen"} 9
    temperature_created{room="kitchen"} 1604597246.08957
    # EOF

You can use this to implement a `/metrics` endpoint in your application
if it is not supported directly. For example, using a raw **httpuv**
server:

``` r
httpuv::runServer(
  "127.0.0.1", 8080,
  list(call = function(req) {
    list(
      status = 200L,
      headers = list("Content-Type" = "text/plain; version=0.0.4"),
      body = render_metrics()
    )
  })
)
```

## Pushgateway Support

Some workloads may not want to run an HTTP server to expose metrics,
especially in the case of short-lived batch jobs. For these cases
metrics can also be manually “pushed” to a Prometheus Pushgateway
instance, though [there are drawbacks to this
approach](https://prometheus.io/docs/practices/pushing/).

`push_to_gateway()` can be used to push metrics, and
`delete_from_gateway()` can be used to clean them up when the workload
is finished:

``` r
push_to_gateway("localhost:9091", job = "openmetrics-readme")
# Some time later...
delete_from_gateway("localhost:9091", job = "openmetrics-readme")
```

## Limitations

  - The Summary metric is not available, as it is not required by the
    spec if a Histogram is available. It is also difficult to implement,
    and [could be considered a design
    error](https://www.robustperception.io/how-does-a-prometheus-summary-work).

  - Most of the default process metrics are absent on non-Linux systems.
    This is basically because they were designed with Linux in mind, and
    very few Prometheus clients support them for anything but Linux.

## License

The package is made available under the terms of the MIT license. See
`LICENSE` for details.


NEWS.md
# openmetrics 0.3.0.9000

* `push_to_gateway()` now has `retry_times` and `httr_config` parameters to
  enable more fine-grained control over the underlying HTTP request -- for
  example, to set timeouts or control SSL behaviour.

* `register_shiny_metrics()` will no longer generate a cryptic error if called
  more than once in a session (#9).

* Counters now have their creation time reset when the total is reset, as
  required by the OpenMetrics standard (#10).

* Minor performance improvements and better error messages.

# openmetrics 0.3.0

## Breaking Changes

* Label names should now be passed as the `labels` parameter when creating a
  metric, instead of being inferred from the remaining named arguments. For
  consistency with other clients, there is no longer a notion of a "default"
  label value. This also makes it possible to have labels like `name`, `help`,
  `type`, and so on that conflict with existing parameter names.

* Relatedly, all labels must now be specified when calling methods on a metric
  (such as `inc()` or `set()`), since there is no default to fall back on.

## Other Improvements

* The package now has its own hex logo.

* Only generate the OpenMetrics format when the (preliminary) OpenMetrics
  content-type is passed in the `Accept` header. Otherwise, fall back on the
  Prometheus format and content-type (#3).

* Pushgateway functions will now `stop()` for HTTP errors instead of merely
  issuing a warning. Clients that can tolerate these errors will likely
  `tryCatch()` them anyway, since `httr::RETRY()` can fail for other reasons.

* Requests made to Shiny apps that return HTTP 401 will no longer include the
  rendered metrics anyway.

* The `"`, `\n`, and `\` characters are now escaped in labels and help text, as
  in the OpenMetrics reference implementation (#1).

* Counter metrics can now be incremented by zero.

* Counter, Gauge, and Histogram metrics can now take an optional `unit`
  parameter. The provided unit *must* match the one in the name of the metric
  itself -- e.g. a `received_bytes` metric must have unit "bytes" (#2).

* Counters and Histograms now have a `_created` child metric, as per the
  OpenMetrics draft specification (#2).

# openmetrics 0.2.0

* Add `push_to_gateway()` and `delete_from_gateway()` for manually pushing
  metrics to a [Prometheus Pushgateway](https://prometheus.io/docs/instrumenting/pushing/)
  instance.

* Histogram metrics now have a built-in `time()` method for collecting
  observations on the duration of an expression.

* The built-in `process_cpu_seconds_total` metric is now correctly reported as
  a Counter instead of a Gauge. This fixes an inconsistency with other
  Prometheus clients that would prevent metrics being used with Pushgateway.

# openmetrics 0.1.1

* Fixes inaccurate CPU usage when using `register_default_metrics()`.
  Previously, this would include the total elapsed time for the process (as
  opposed to the CPU time that was actually used).

# openmetrics 0.1.0

* Initial release. **openmetrics** is an opinionated Prometheus and OpenMetrics
  client for R.

* Automatically instrument Plumber apps with `register_plumber_metrics()` and
  Shiny apps with `register_shiny_metrics()`.

* Collect default metrics with `register_default_metrics()`.

* Create custom metrics with `counter_metric()`, `gauge_metric()`, and
  `histogram_metric()` and custom registries with `registry()`.

* Write metrics in the OpenMetrics/Prometheus text format to any location with
  `render_metrics()`.


tests/testthat.R
library(testthat)
library(openmetrics)

test_check("openmetrics")


tests/testthat/test-labels.R
testthat::test_that("Labels work as expected", {
  valid <- list(colour = "red", month = "aug")
  available <- c("colour", "month")

  # Ingesting valid labels.
  testthat::expect_equal(parse_labels(list()), NULL)
  testthat::expect_equal(parse_labels(valid), valid)

  # Catching invalid labels.
  testthat::expect_error(
    parse_labels(list("red", "aug")),
    regexp = "All labels must be named"
  )
  testthat::expect_error(
    parse_labels(list(colour = "red", "aug")),
    regexp = "All labels must be named"
  )
  testthat::expect_error(
    parse_labels(list(colour = c("red", "blue"))),
    regexp = "All labels must be strings"
  )
  testthat::expect_error(
    parse_labels(list(colour = 1)),
    regexp = "All labels must be strings"
  )

  # Validating labels.
  testthat::expect_equal(validate_labels(valid, available), valid)
  testthat::expect_error(
    validate_labels(NULL, available),
    regexp = "One or more missing metric labels"
  )
  testthat::expect_error(
    validate_labels(list(colour = "yellow"), available),
    regexp = "One or more missing metric labels"
  )

  # Encoding.
  testthat::expect_equal(encode_labels(valid), 'colour="red",month="aug"')
})

testthat::test_that("Escaping works as expected", {
  misbehaved <- list(unescaped = "cats\n a\\nd \"dogs\"\n", fine = "text")
  testthat::expect_equal(
    encode_labels(misbehaved),
    'unescaped="cats\\n a\\\\nd \\\"dogs\\\"\\n",fine="text"'
  )
})


R/shiny.R
#' Metrics for Shiny Applications
#'
#' @description
#'
#' Automatically wrap a Shiny app, adding metrics for the current session count
#' and the duration of reactive flushes, and then expose them on a `/metrics`
#' endpoint.
#'
#' The endpoint will check the `METRICS_HTTP_AUTHORIZATION` environment
#' variable, and if present will use it as the expected
#' [`Authorization`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization)
#' header of the request to the `/metrics` endpoint. This can be used to
#' implement basic HTTP authentication for access to runtime metrics.
#'
#' @param app An object created with [shiny::shinyApp()].
#' @param registry A `Registry` object. See [registry()].
#' @return A modified Shiny app object.
#'
#' @export
register_shiny_metrics <- function(app, registry = openmetrics::global_registry()) {
  stopifnot(inherits(app, "shiny.appobj"))

  # Grab Shiny internals. This is a hack to avoid triggering R CMD check warnings
  # for the use of shiny:::*.
  httpResponse <- get0("httpResponse", envir = asNamespace("shiny"))
  handlerManager <- get0("handlerManager", envir = asNamespace("shiny"))
  if (is.null(httpResponse) || is.null(handlerManager)) {
    stop("This version of Shiny is not supported.")
  }

  session_count <- gauge_metric(
    "shiny_sessions", "Running count of open Shiny sessions.",
    registry = registry
  )
  duration <- histogram_metric(
    "shiny_flush_duration_seconds", "Duration of reactive flushes, in seconds.",
    # This seems to be in the right range for most flushes.
    buckets = c(0.0001, 0.00025, 0.0005, 0.001, 0.0025, 0.01),
    registry = registry
  )

  flush_timestamp <- NULL
  server_fun <- app$serverFuncSource()
  app$serverFuncSource <- function() {
    function(input, output, session) {
      session_count$inc() # Up the session count.

      # Collect metrics on various Shiny events.
      shiny::onSessionEnded(function() {
        session_count$dec()
      })
      shiny::onFlush(function() {
        flush_timestamp <<- unclass(Sys.time())
      }, once = FALSE)
      shiny::onFlushed(function() {
        duration$observe(unclass(Sys.time()) - flush_timestamp)
      }, once = FALSE)

      server_fun(input, output, session)
    }
  }

  metrics_handler <- function(req) {
    if (req$PATH_INFO != "/metrics") {
      NULL
    } else {
      # Check for authorization, if set.
      auth_header <- Sys.getenv("METRICS_HTTP_AUTHORIZATION")
      if (nchar(auth_header) > 0 && (is.null(req$HTTP_AUTHORIZATION) ||
                                     req$HTTP_AUTHORIZATION != auth_header)) {
        httpResponse(
          content_type = "text/plain", status = 401L,
          headers = list(
            "Content-Type" = "text/plain",
            "WWW-Authenticate" = "Basic realm=\"Runtime metrics\""
          ),
          content = "Unauthorized"
        )
      } else {
        if (!is.null(req$HTTP_ACCEPT) &&
            grepl("application/openmetrics-text", req$HTTP_ACCEPT, fixed = TRUE)) {
          httpResponse(
            content_type = .content_type,
            content = registry$render_all()
          )
        } else {
          httpResponse(
            content_type = .legacy_content_type,
            content = registry$render_all(format = "legacy")
          )
        }
      }
    }
  }

  # Prevent 'Key in use' errors when this function is called more than once.
  handlerManager$removeHandler("/metrics")
  handlerManager$addHandler(metrics_handler, "/metrics")
  app
}


R/metrics.R
#' Metrics
#'
#' @description
#'
#' A metric is a measure which can be aggregated into a time series, and comes
#' in one of three types: counters, gauges, and histograms.
#'
#' Metrics must have a unique name.
#'
#' @param name The name of the metric.
#' @param help A brief, one-sentence explanation of the metric's meaning.
#' @param labels A vector of label names for the metric.
#' @param unit An optional unit for the metric, e.g. \code{"seconds"}. Must
#'   match the metric name.
#' @param ... For backward compatibility, otherwise ignored.
#' @param registry Where to register the metric for later retrieval.
#'
#' @return An object with methods to manipulate the metric. See details.
#'
#' @details
#'
#' All metric objects have a `reset()` method that reverts the underlying value
#' (or values) to zero, an `unregister()` method that removes them from the
#' registry they were created in, and a `render()` method that writes a
#' representation of the metric in the text-based [OpenMetrics
#' format](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format).
#' Normally, [render_metrics()] is used instead.
#'
#' In addition, various metrics have their own methods:
#'
#' * `inc(by = 1, ...)`: Increments the metric by some positive number,
#'   defaulting to 1. Further parameters are interpreted as labels. Available
#'   for counters and gauges.
#'
#' * `dec(by = 1, ...)`: Decrements the metric by some number, defaulting to 1.
#'   Further parameters are interpreted as labels. Available for gauges.
#'
#' * `set(value, ...)`: Sets the metric to some number. Further parameters are
#'   interpreted as labels. Available for gauges.
#'
#' * `set_to_current_time(...)`: Sets the metric to the current time, in seconds
#'   from the Unix epoch. Further parameters are interpreted as labels.
#'   Available for gauges.
#'
#' * `observe(value, ...)`: Records an observation of some number. Further
#'   parameters are interpreted as labels. Available for histograms.
#'
#' * `time(expr, ...)`: Records an observation for the time elapsed evaluating
#'   `expr`, in seconds. Further parameters are interpreted as labels.
#'   Available for histograms.
#'
#' @examples
#' meows <- counter_metric("meows", "Heard around the house.", labels = "cat")
#' meows$inc(cat = "Shamus") # Count one meow from Shamus.
#' meows$inc(3, cat = "Unknown") # Count three meows of unknown origin.
#' meows$render()
#'
#' thermostat <- gauge_metric("thermostat", "Thermostat display.")
#' thermostat$set(21.3) # Read from the display...
#' thermostat$dec(2) # ... and then turn it down 2 degrees.
#' thermostat$render()
#'
#' temperature <- histogram_metric(
#'   "temperature", "Ambient room temperature measurements.",
#'   buckets = c(10, 15, 20, 22, 25), labels = "room"
#' )
#' set.seed(9090)
#' # Simulate taking ambient temperature samples.
#' for (measure in rnorm(20, mean = 21.5)) {
#'   temperature$observe(measure, room = sample(c("kitchen", "bathroom"), 1))
#' }
#' temperature$render()
#'
#' @seealso The official documentation on [Metric Types](https://prometheus.io/docs/concepts/metric_types/).
#' @name metrics
#' @export
counter_metric <- function(name, help, labels = character(), ..., unit = NULL,
                           registry = global_registry()) {
  existing <- registry$metric(name, type = "counter")
  if (is.null(existing)) {
    # Backwards compatibility with passing labels as arguments.
    labels <- union(labels, names(list(...)))
    Counter$new(
      name = name, help = help, labels = labels, unit = unit,
      registry = registry
    )
  } else {
    existing
  }
}

#' @rdname metrics
#' @export
gauge_metric <- function(name, help, labels = character(), ..., unit = NULL,
                         registry = global_registry()) {
  existing <- registry$metric(name, type = "gauge")
  if (is.null(existing)) {
    # Backwards compatibility with passing labels as arguments.
    labels <- union(labels, names(list(...)))
    Gauge$new(
      name = name, help = help, labels = labels, unit = unit,
      registry = registry
    )
  } else {
    existing
  }
}

#' @param buckets A sequence of buckets to bin observations into. Defaults to
#'   Prometheus's suggested buckets, which are a good fit for measuring
#'   user-visible latency in seconds (e.g. for web services).
#' @rdname metrics
#' @export
histogram_metric <- function(name, help, buckets = c(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10),
                             labels = character(), ..., unit = NULL,
                             registry = global_registry()) {
  existing <- registry$metric(name, type = "histogram")
  if (is.null(existing)) {
    # Backwards compatibility with passing labels as arguments.
    labels <- union(labels, names(list(...)))
    Histogram$new(
      name = name, help = help, buckets = buckets, labels = labels, unit = unit,
      registry = registry
    )
  } else {
    existing
  }
}

Metric <- R6::R6Class(
  "Metric",
  public = list(
    initialize = function(name, help, type = "gauge", labels = character(),
                          unit = NULL, registry = global_registry()) {
      if (!grepl("^[a-zA-Z_:][a-zA-Z0-9_:]*$", name)) {
        stop("Invalid metric name: '", name, "'.")
      }
      private$name <- name
      private$help <- escape_text(help)
      private$type <- type
      private$unit <- unit
      private$registry <- registry
      if (length(labels) > 0) {
        # This is so we don't have to refactor the NULL checks.
        private$labels <- labels
      }

      # We only need to check spec validity for label names once.
      valid <- grepl("^[a-zA-Z_][a-zA-Z0-9_]*$", private$labels)
      if (!all(valid)) {
        stop(
          "One or more invalid metric labels: '",
          paste(private$labels[!valid], collapse = "', '"), "'."
        )
      }

      if (!is.null(private$unit) &&
          !grepl(sprintf("_%s$", private$unit), private$name)) {
        stop("Metric name does not match unit.")
      }

      registry$register(name = name, type = type, self)
    },

    unregister = function() {
      private$registry$unregister(name = private$name, type = private$type)
    }
  ),
  private = list(
    name = NULL, help = NULL, type = NULL, labels = NULL, registry = NULL,
    unit = NULL,

    header = function(name = private$name, type = private$type) {
      if (!is.null(private$unit)) {
        sprintf(
          "# HELP %s %s\n# TYPE %s %s\n# UNIT %s %s", name, private$help, name,
          type, name, private$unit
        )
      } else {
        sprintf(
          "# HELP %s %s\n# TYPE %s %s", name, private$help, name, type
        )
      }
    }
  )
)

Counter <- R6::R6Class(
  "Counter", inherit = Metric,
  public = list(
    initialize = function(name, help, labels = character(), unit = NULL,
                          registry = global_registry()) {
      # Compatibility with OpenMetrics.
      name <- gsub("_total$", "", name)
      super$initialize(
        name, help, type = "counter", labels = labels, unit = unit,
        registry = registry
      )
      if (is.null(private$labels)) {
        # Fast version, just a value.
        private$value <- 0
        private$created <- unclass(Sys.time())
      } else {
        private$value <- new.env(parent = emptyenv())
        private$created <- new.env(parent = emptyenv())
      }
    },

    render = function(format = "openmetrics") {
      # Compatibility with OpenMetrics requires that metric names include
      # _total but help/type text do not. However, some existing tools will
      # barf on this input, notably the Prometheus Pushgateway, so it must be
      # possible to circumvent.
      #
      # In addition, when in legacy mode, make _created a separate metric with
      # no HELP or TYPE info. This is what the Python reference client does.
      if (format == "openmetrics") {
        header <- private$header()
        if (is.null(private$labels)) {
          sprintf(
            "%s\n%s_total %s\n%s_created %s\n", header, private$name,
            private$value, private$name, private$created
          )
        } else {
          entries <- vapply(ls(private$value), function(key) {
            sprintf(
              "%s_total{%s} %s\n%s_created{%s} %s", private$name, key,
              private$value[[key]], private$name, key, private$created[[key]]
            )
          }, character(1))
          sprintf("%s\n%s\n", header, paste(entries, collapse = "\n"))
        }
      } else {
        name <- sprintf("%s_total", private$name)
        header <- private$header(name = name)
        if (is.null(private$labels)) {
          sprintf(
            "%s\n%s %s\n# TYPE %s_created gauge\n%s_created %s\n", header, name,
            private$value, name, name, private$created
          )
        } else {
          entries <- vapply(ls(private$value), function(key) {
            sprintf("%s{%s} %s", name, key, private$value[[key]])
          }, character(1))
          created <- vapply(ls(private$created), function(key) {
            sprintf("%s_created{%s} %s", name, key, private$created[[key]])
          }, character(1))
          sprintf(
            "%s\n%s\n# TYPE %s_created gauge\n%s\n", header,
            paste(entries, collapse = "\n"), name,
            paste(created, collapse = "\n")
          )
        }
      }
    },

    inc = function(by = 1, ...) {
      if (by < 0) {
        stop(
          "Counter metrics can only be incremented by a positive number.",
          call. = FALSE
        )
      }
      if (is.null(private$labels)) {
        private$value <- private$value + by
      } else {
        key <- encode_labels(validate_labels(list(...), private$labels))
        value <- private$value[[key]]
        if (is.null(value)) {
          private$created[[key]] <- unclass(Sys.time())
          private$value[[key]] <- by
        } else {
          private$value[[key]] <- value + by
        }
      }
    },

    reset = function() {
      if (is.null(private$labels)) {
        private$value <- 0
      } else {
        private$created <- new.env(parent = emptyenv())
        private$value <- new.env(parent = emptyenv())
      }
    }
  ),
  private = list(
    value = NULL, created = NULL
  )
)

Gauge <- R6::R6Class(
  "Gauge", inherit = Metric,
  public = list(
    initialize = function(name, help, labels = character(), unit = NULL,
                          registry = global_registry()) {
      super$initialize(
        name, help, type = "gauge", labels = labels, unit = unit,
        registry = registry
      )
      if (is.null(private$labels)) {
        # Fast version, just a value.
        private$value <- 0
      } else {
        private$value <- new.env(parent = emptyenv())
      }
    },

    render = function(format = "openmetrics") {
      if (is.null(private$labels)) {
        sprintf(
          "%s\n%s %s\n", private$header(), private$name, private$value
        )
      } else {
        entries <- vapply(ls(private$value), function(key) {
          sprintf("%s{%s} %s", private$name, key, private$value[[key]])
        }, character(1))
        sprintf(
          "%s\n%s\n", private$header(), paste(entries, collapse = "\n")
        )
      }
    },

    inc = function(by = 1, ...) {
      if (is.null(private$labels)) {
        private$value <- private$value + by
      } else {
        key <- encode_labels(validate_labels(list(...), private$labels))
        value <- private$value[[key]]
        if (is.null(value)) {
          private$value[[key]] <- by
        } else {
          private$value[[key]] <- value + by
        }
      }
    },

    dec = function(by = 1, ...) {
      if (is.null(private$labels)) {
        private$value <- private$value - by
      } else {
        key <- encode_labels(validate_labels(list(...), private$labels))
        value <- private$value[[key]]
        if (is.null(value)) {
          private$value[[key]] <- -by
        } else {
          private$value[[key]] <- value - by
        }
      }
    },

    set = function(value, ...) {
      if (!is.numeric(value)) {
        stop("Gauge metrics can only be set to a number.", call. = FALSE)
      }
      if (is.null(private$labels)) {
        private$value <- value
      } else {
        key <- encode_labels(validate_labels(list(...), private$labels))
        private$value[[key]] <- value
      }
    },

    set_to_current_time = function(...) {
      self$set(value = unclass(Sys.time(), ...))
    },

    reset = function() {
      if (is.null(private$labels)) {
        private$value <- 0
      } else {
        private$value <- new.env(parent = emptyenv())
      }
    }
  ),
  private = list(
    value = NULL
  )
)

Histogram <- R6::R6Class(
  "Histogram", inherit = Metric,
  public = list(
    initialize = function(name, help, buckets = c(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10),
                          labels = character(), unit = NULL,
                          registry = global_registry()) {
      super$initialize(
        name, help, type = "histogram", labels = labels, unit = unit,
        registry = registry
      )
      buckets <- sort(buckets)
      private$buckets <- c(buckets, Inf)
      # Note: Buckets without a .0 seem to need them to satisfy the parser, but
      # otherwise the usual numeric formatting seems to be acceptable.
      buckets_str <- as.character(buckets)
      have_digits <- abs(buckets - trunc(buckets)) > 0
      buckets_str[!have_digits] <- sprintf("%.1f", buckets[!have_digits])
      private$le <- c(buckets_str, "+Inf")
      self$reset()
      if (is.null(private$labels)) {
        private$created <- unclass(Sys.time())
      } else {
        private$created <- new.env(parent = emptyenv())
      }
    },

    render = function(format = "openmetrics") {
      if (is.null(private$labels)) {
        buckets <- paste0(
          private$name, "_bucket{le=\"", private$le, "\"} ", private$dist
        )
        sums <- sprintf("%s_sum %s", private$name, private$sum)
        count <- sprintf("%s_count %s", private$name, private$count)
        created <- sprintf("%s_created %s", private$name, private$created)
        if (format == "openmetrics") {
          sprintf(
            "%s\n%s\n", private$header(),
            paste(c(buckets, sums, count, created), collapse = "\n")
          )
        } else {
          created_type <- sprintf("#TYPE %s_created gauge", private$name)
          sprintf(
            "%s\n%s\n", private$header(), paste(
              c(buckets, sums, count, created_type, created), collapse = "\n"
            )
          )
        }
      } else {
        # Matching labels need to be printed together to satisfy the OpenMetrics
        # parser.
        keys <- ls(private$dist)
        if (length(keys) == 0) {
          return("")
        }
        blocks <- character(length(keys))
        # For legacy formatting, keep _created under its own metric.
        if (format == "openmetrics") {
          cblocks <- character()
        } else {
          cblocks <- character(length(keys) + 1)
          cblocks[1] <- sprintf("# TYPE %s_created gauge", private$name)
        }
        for (i in 1:length(keys)) {
          key <- keys[i]
          buckets <- paste0(
            private$name, "_bucket{", key, ",le=\"", private$le, "\"} ",
            private$dist[[key]]
          )
          sums <- sprintf("%s_sum{%s} %s", private$name, key, private$sum[[key]])
          count <- sprintf(
            "%s_count{%s} %s", private$name, key, private$count[[key]]
          )

          if (format == "openmetrics") {
            created <- sprintf(
              "%s_created{%s} %s", private$name, key, private$created[[key]]
            )
            blocks[i] <- paste(c(buckets, sums, count, created), collapse = "\n")
          } else {
            blocks[i] <- paste(c(buckets, sums, count), collapse = "\n")
            cblocks[i + 1] <- sprintf(
              "%s_created{%s} %s", private$name, key, private$created[[key]]
            )
          }
        }
        sprintf(
          "%s\n%s\n", private$header(),
          paste(c(blocks, cblocks), collapse = "\n")
        )
      }
    },

    observe = function(value, ...) {
      if (!is.numeric(value)) {
        stop("Histogram metrics can only observe a number.", call. = FALSE)
      }
      dist <- as.integer(value <= private$buckets)

      # Update the running sum and count values.
      if (is.null(private$labels)) {
        private$dist <- private$dist + dist
        private$count <- private$count + 1
        private$sum <- private$sum + value
      } else {
        key <- encode_labels(validate_labels(list(...), private$labels))
        current <- private$dist[[key]]
        # We know that if one entry is missing they all are.
        if (is.null(current)) {
          private$created[[key]] <- unclass(Sys.time())
          private$dist[[key]] <- dist
          private$count[[key]] <- 1
          private$sum[[key]] <- value
        } else {
          private$dist[[key]] <- private$dist[[key]] + dist
          private$count[[key]] <- private$count[[key]] + 1
          private$sum[[key]] <- private$sum[[key]] + value
        }
      }
    },

    time = function(expr, ...) {
      start <- Sys.time()
      expr
      elapsed <- unclass(difftime(Sys.time(), start, units = "secs"))
      self$observe(elapsed, ...)
      elapsed
    },

    reset = function() {
      if (is.null(private$labels)) {
        private$sum <- 0
        private$count <- 0
        private$dist <- rep(0, times = length(private$buckets))
      } else {
        private$dist <- new.env(parent = emptyenv())
        private$sum <- new.env(parent = emptyenv())
        private$count <- new.env(parent = emptyenv())
      }
    }
  ),
  private = list(
    buckets = numeric(), le = character(), dist = NULL, sum = NULL,
    count = NULL, created = NULL
  )
)

parse_labels <- function(labels) {
  if (length(labels) == 0) {
    return(NULL)
  }
  if (is.null(names(labels)) || any(nchar(names(labels)) == 0)) {
    stop("All labels must be named.", call. = FALSE)
  }
  strings <- vapply(
    labels, function(x) is.character(x) && length(x) == 1, logical(1)
  )
  if (!all(strings)) {
    stop("All labels must be strings.", call. = FALSE)
  }
  labels
}

validate_labels <- function(labels, available) {
  labels <- parse_labels(labels)
  present <- available %in% names(labels)
  if (!all(present)) {
    missing <- paste0(available[!present], collapse = "', '")
    stop("One or more missing metric labels: '", missing,  "'.")
  }
  # TODO: Is it wrong to silently drop additional labels?
  labels[names(labels) %in% available]
}

encode_labels <- function(labels) {
  paste0(names(labels), "=\"", escape_text(labels), "\"", collapse = ",")
}

escape_text <- function(text) {
  text <- gsub("\\", "\\\\", text, fixed = TRUE)
  text <- gsub("\n", "\\n", text, fixed = TRUE)
  text <- gsub("\"", "\\\"", text, fixed = TRUE)
  text
}

.content_type <- "application/openmetrics-text;version=0.0.1;charset=utf-8"
.legacy_content_type <- "text/plain;version=0.0.4;charset=utf-8"
